add containerd runtime

tianyax opened the pull request:
# Description

For #1232 

add containerd runtime support.

```bash
# Install Dapr
>>> dapr init --container-runtime containerd --network mynet
⌛  Making the jump to hyperspace...
ℹ️  Container images will be pulled from Docker Hub
ℹ️  Installing runtime version 1.10.2
↖  Downloading binaries and setting up components...
✅  Downloading binaries and setting up components...
✅  Downloaded binaries and completed components set up.
ℹ️  daprd binary has been installed to /home/x.linux/.dapr/bin.
ℹ️  dapr_placement_mynet container is running.
ℹ️  dapr_redis_mynet container is running.
ℹ️  dapr_zipkin_mynet container is running.
ℹ️  Use `nerdctl ps` to check running containers.
✅  Success! Dapr is up and running. To get started, go here: https://aka.ms/dapr-getting-started

# Show all the containers
>>> nerdctl ps
CONTAINER ID    IMAGE                                 COMMAND                   CREATED               STATUS    PORTS    NAMES
55f8cd7ad1ad    docker.io/daprio/dapr:1.10.2          "./placement"             About a minute ago    Up                 dapr_placement_mynet
6c4a54236073    docker.io/openzipkin/zipkin:latest    "start-zipkin"            About a minute ago    Up                 dapr_zipkin_mynet
b94b3d45d466    docker.io/library/redis:6             "docker-entrypoint.s…"    About a minute ago    Up                 dapr_redis_mynet

# Uninstall dapr
>>> dapr uninstall --container-runtime containerd --network mynet 

```

## Issue reference

Please reference the issue this PR will close: #1232 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation

tianyax committed:
support nerdctl+containerd runtime (#1232)

Signed-off-by: tianya <tianyax@outlook.com>

tianyax committed:
add containerd runtime validity check and fix test errors

Signed-off-by: tianya <tianyax@outlook.com>

tianyax committed:
reduce unnecessary container runtime availability checks

Signed-off-by: tianya <tianyax@outlook.com>

tianyax committed:
fixed test failed on macos

Signed-off-by: tianya <tianyax@outlook.com>

pravinpushkar commented on cmd/init.go:
```diff
@@ -148,7 +148,7 @@ dapr init --runtime-path <path-to-install-directory>
 
 			if !utils.IsValidContainerRuntime(containerRuntime) {
-				print.FailureStatusEvent(os.Stdout, "Invalid container runtime. Supported values are docker and podman.")
+				print.FailureStatusEvent(os.Stdout, "Invalid container runtime. Supported values are docker and podman and containerd.")
```
```suggestion
				print.FailureStatusEvent(os.Stdout, "Invalid container runtime. Supported values are docker, podman and containerd.")
```

pravinpushkar commented on cmd/init.go:
```diff
@@ -194,7 +194,7 @@ func init() {
 	InitCmd.Flags().StringArrayVar(&values, "set", []string{}, "set values on the command line (can specify multiple or separate values with commas: key1=val1,key2=val2)")
 	InitCmd.Flags().String("image-registry", "", "Custom/private docker image repository URL")
-	InitCmd.Flags().StringVarP(&containerRuntime, "container-runtime", "", "docker", "The container runtime to use. Supported values are docker (default) and podman")
+	InitCmd.Flags().StringVarP(&containerRuntime, "container-runtime", "", "docker", "The container runtime to use. Supported values are docker (default) and podman and containerd")
```
```suggestion
	InitCmd.Flags().StringVarP(&containerRuntime, "container-runtime", "", "docker", "The container runtime to use. Supported values are docker (default),  podman and containerd")
```

pravinpushkar commented on cmd/uninstall.go:
```diff
@@ -69,7 +69,7 @@ dapr uninstall --runtime-path <path-to-install-directory>
 		} else {
 			if !utils.IsValidContainerRuntime(uninstallContainerRuntime) {
-				print.FailureStatusEvent(os.Stdout, "Invalid container runtime. Supported values are docker and podman.")
+				print.FailureStatusEvent(os.Stdout, "Invalid container runtime. Supported values are docker and podman and containerd.")
```
```suggestion
				print.FailureStatusEvent(os.Stdout, "Invalid container runtime. Supported values are docker, podman and containerd.")
```

pravinpushkar commented on cmd/uninstall.go:
```diff
@@ -92,6 +92,6 @@ func init() {
 	UninstallCmd.Flags().StringVarP(&uninstallNamespace, "namespace", "n", "dapr-system", "The Kubernetes namespace to uninstall Dapr from")
 	UninstallCmd.Flags().BoolP("help", "h", false, "Print this help message")
-	UninstallCmd.Flags().StringVarP(&uninstallContainerRuntime, "container-runtime", "", "docker", "The container runtime to use. Supported values are docker (default) and podman")
+	UninstallCmd.Flags().StringVarP(&uninstallContainerRuntime, "container-runtime", "", "docker", "The container runtime to use. Supported values are docker (default) and podman and containerd")
```
```suggestion
	UninstallCmd.Flags().StringVarP(&uninstallContainerRuntime, "container-runtime", "", "docker", "The container runtime to use. Supported values are docker (default),  podman and containerd")
```

pravinpushkar commented on utils/utils.go:
```diff
@@ -38,25 +39,37 @@ import (
 func GetContainerRuntimeCmd(containerRuntime string) string {
+	// containerd runtime use nerdctl tool.
+	if IsValidRuntimeOnMacos(containerRuntime) {
+		return "nerdctl"
```
putting this also as constant would be better ?

tianyax replied:
👍

tianyax committed:
update help message

Co-authored-by: Pravin Pushkar <ppushkar@microsoft.com>
Signed-off-by: tianya <12004412+tianyax@users.noreply.github.com>

tianyax committed:
use the string nerdctl as a constant

Signed-off-by: tianya <tianyax@outlook.com>

pravinpushkar commented on utils/utils.go:
```diff
@@ -38,25 +39,38 @@ import (
 )
 
+var IsValidRuntimeOnMacos = func(containerRuntime string) bool {
+	containerRuntime = strings.TrimSpace(containerRuntime)
```
See this comment- https://github.com/dapr/cli/issues/1232#issuecomment-1447590195

tianyax committed:
revert containerd support on macos

Signed-off-by: tianya <tianyax@outlook.com>

tianyax committed:
update README

Signed-off-by: tianya <tianyax@outlook.com>

pravinpushkar commented:
@tianyax Were you able to verify `dapr init` with `containerd` on `Macos`?

pravinpushkar committed:
Merge branch 'master' into containerd-runtime

tianyax commented:
> @tianyax Were you able to verify running `dapr init` with `containerd` on `Macos`?

@pravinpushkar  Okay, I'll update the workflows later.

pravinpushkar commented:
> > @tianyax Were you able to verify running `dapr init` with `containerd` on `Macos`?
> 
> @pravinpushkar Okay, I'll update the workflows later.

I meant locally.

However, should we try updating the workflow for Mac using `containerd` or have both `podman` and `containerd` @mukundansundar

mukundansundar commented:
> > > @tianyax Were you able to verify running `dapr init` with `containerd` on `Macos`?
> 
> > 
> 
> > @pravinpushkar Okay, I'll update the workflows later.
> 
> 
> 
> I meant locally.
> 
> 
> 
> However, should we try updating the workflow for Mac using `containerd` or have both `podman` and `containerd` @mukundansundar 

If possible both.

tianyax commented:
> > > @tianyax Were you able to verify running `dapr init` with `containerd` on `Macos`?
> > 
> > 
> > @pravinpushkar Okay, I'll update the workflows later.
> 
> I meant locally.
> 
> However, should we try updating the workflow for Mac using `containerd` or have both `podman` and `containerd` @mukundansundar

@pravinpushkar I have already verified it on my mac, including `lima`, `colima` and `rancher-desktop`.
This is the result after I cleared the environment and re-run with `colima` as an example.
<img width="1920" alt="image" src="https://user-images.githubusercontent.com/12004412/222345939-d1cfcd6b-1b0d-45c7-a77a-bfdb0dce6307.png">

pravinpushkar commented:
@tianyax Thanks for verifying locally. Please resolve the conflicts in the PR .

tianyax committed:
Merge remote-tracking branch 'master' into containerd-runtime

tianyax committed:
Merge branch 'master' into containerd-runtime

tianyax committed:
Merge branch 'master' into containerd-runtime

Signed-off-by: tianya <12004412+tianyax@users.noreply.github.com>

tianyax committed:
fixed args missing

Signed-off-by: tianya <tianyax@outlook.com>

mukundansundar committed:
Merge branch 'master' into containerd-runtime

pravinpushkar commented:
@tianyax There is small merge conflict that needs to be resolved.

tianyax committed:
Merge branch 'master' into containerd-runtime

Signed-off-by: tianya <tianyax@outlook.com>

tianyax commented:
> @tianyax There is small merge conflict that needs to be resolved.

@pravinpushkar resolved.

shubham1172 commented on README.md:
```diff
@@ -11,14 +11,49 @@ The Dapr CLI allows you to setup Dapr on your local dev machine or on a Kubernet
+On default, during initialization the Dapr CLI will install the Dapr binaries as well as setup a developer environment to help you get started easily with Dapr. This environment uses `Docker`, `Podman` or `Containerd` containers, therefore container runtime needs to be installed. If you prefer to run Dapr without this environment and without any container dependencies, after installation of the C ...
 
-Note, if you are a new user, it is strongly recommended to install Docker and use the regular init command.
+#### Note, if you are a new user, it is strongly recommended to install Docker and use the regular init command.
```
nit: can we use a quote instead of header?
```suggestion
>__Note, if you are a new user, it is strongly recommended to install Docker and use the regular init command.__
```

shubham1172 commented on README.md:
```diff
@@ -11,14 +11,49 @@ The Dapr CLI allows you to setup Dapr on your local dev machine or on a Kubernet
 
 >__Note: On Windows, Docker must be running in Linux Containers mode__
 
+#### If you want to use podman as a runtime, then please refer to the installation.
+
+* Install [Podman](https://podman-desktop.io/docs/Installation)
+
+#### If you want to use containerd as a runtime, then please refer to the installation.
```
This can be simpler.

#### Choosing a container runtime
##### Docker (recommended for new users)
Docker instructions
##### Podman
Podman instructions
##### Containerd
Containerd instructions

tianyax replied:
Okay, thanks very much!

shubham1172 commented on cmd/uninstall.go:
```diff
@@ -52,11 +52,11 @@ dapr uninstall -k
 		viper.BindPFlag("install-path", cmd.Flags().Lookup("install-path"))
 	},
-	Run: func(cmd *cobra.Command, args []string) {
+	Run: func(*cobra.Command, []string) {
```
```suggestion
	Run: func(*cobra.Command, _ []string) {
```

tianyax replied:
Is that what you mean?
```go
func(cmd *cobra.Command, _ []string) {}
```
Here the parameter is not used, I think it might be better to follow the @go language style if possible.😄️

shubham1172 replied:
Yes, we should agree on one single convention in this codebase. The other place used `_` (install.go), so recommended the same. 

Ok with using this convention - `Run: func(*cobra.Command, []string) {` in all places.

shubham1172 commented on pkg/standalone/standalone.go:
```diff
@@ -161,9 +161,8 @@ func Init(runtimeVersion, dashboardVersion string, dockerNetwork string, slimMod
-		// If --slim installation is not requested, check if docker is installed.
-		conatinerRuntimeAvailable := utils.IsDockerInstalled() || utils.IsPodmanInstalled()
-		if !conatinerRuntimeAvailable {
+		// If --slim installation is not requested, check if docker, podman or containerd is installed.
```
```suggestion
		// If --slim installation is not requested, check if the container runtime is installed.
```

shubham1172 commented on utils/utils.go:
```diff
@@ -38,30 +38,38 @@ import (
 
 // GetContainerRuntimeCmd returns a valid container runtime to be used by CLI operations.
-// If the input is a valid container runtime, it is returned as is.
+// If the input is a valid container runtime, it is returned client tool.
```
```suggestion
// If the input is a valid container runtime, the client tool is returned.
```

shubham1172 commented on utils/utils.go:
```diff
@@ -188,6 +196,27 @@ func IsPodmanInstalled() bool {
 
+// IsContainerdInstalled checks whether nerdctl and containerd is installed/running.
+func IsContainerdInstalled() bool {
+	if _, err := RunCmdAndWait("nerdctl", "info"); err != nil {
```
Is nerdctl installation enough to check? Don't we need to check the actual containerd daemon?

tianyax replied:
Yes, `nerdctl info` will check `containerd` daemon status, so I need to use `RunCmdAndWait` to wait for the output result. In my mac (using `colima`), this error occurs: `FATA[0000] colima is not running`; and linux will print like this: `FATA[0000] cannot access containerd socket "/run/containerd/containerd.sock": no such file or directory`

shubham1172 reviewed (changes requested):
Can we also add e2e tests with containerd?

tianyax commented:
> Can we also add e2e tests with containerd?

@shubham1172 `containerd` just provides a runtime, the e2e test cases can be equivalent to docker, I think we can add the `containerd` job to the `self-host-e2e` workflow to do e2e tests, if that's what you mean, I try to update it.

shubham1172 commented:
> I think we can add the containerd job to the self-host-e2e workflow to do e2e tests, if that's what you mean, I try to update it.

@tianyax, absolutely!

tianyax committed:
Apply suggestions from code review

Co-authored-by: Shubham Sharma <shubhash@microsoft.com>
Signed-off-by: tianya <tianyax@outlook.com>

tianyax committed:
update runtime installation in README

Signed-off-by: tianya <tianyax@outlook.com>

tianyax committed:
add e2e tests with containerd

Signed-off-by: tianya <tianyax@outlook.com>

shubham1172 committed:
Merge branch 'master' into containerd-runtime

pravinpushkar commented:
Yeah, better to run the test locally to debug it. Also, any reason why enable it only for macos and not windows and linux ?

tianyax commented:
> > @shubham1172 @pravinpushkar According to shubham1172's suggestion I modified `README` and added e2e tests for `containerd` runtime. But the [job](https://github.com/dapr/cli/actions/runs/4524712614/jobs/7968721746) timed out and I am looking for the cause, I suspect that the command `dapr init ...` is invoked in go that does not get the result and make the program waiting. No problem using `dapr init ...` locally
> 
> Hi @tianyax, are you able to run the e2e test locally?

@shubham1172 Yes, I have no problem running e2e tests locally, but my `colima` environment is not freshly installed, it has been running for a while, I will clean it and do a retest.
![image](https://user-images.githubusercontent.com/12004412/228172469-e28c015f-8e53-42a1-baab-d154f8e46a92.png)

tianyax commented:
> Yeah, better to run the test locally to debug it. Also, any reason why enable it only for macos and not windows and linux ?

@pravinpushkar Yes, I will be debug it locally. About `why enable it only for macos`, I saw that other runtimes just chose a platform to do the test. I think there may be some reasons for this. I am using macos now, so I chose macos to do the test. After solving this problem I will add. 😄️

pravinpushkar commented:
> @pravinpushkar Yes, I will be debug it locally. About `why enable it only for macos`, I saw that other runtimes just chose a platform to do the test. I think there may be some reasons for this. I am using macos now, so I chose macos to do the test. After solving this problem I will add. 😄️

Basically we recently added `Podman` support and macos runner has some issues with docker and windows runner has some difficulties with `Podman`, thats why the current setup looks like this. I was just thinking if `containerd` has good support for all then we can add for all os and exclude what is not supported.

tianyax commented:
> @shubham1172 @pravinpushkar According to shubham1172's suggestion I modified `README` and added e2e tests for `containerd` runtime. But the [job](https://github.com/dapr/cli/actions/runs/4524712614/jobs/7968721746) timed out and I am looking for the cause, I suspect that the command `dapr init ...` is invoked in go that does not get the result and make the program waiting. No problem using `dapr init ...` locally

@shubham1172 @pravinpushkar The problem is that the `utils.RunCmdAndWait` method causes blocking.
```go
func RunCmdAndWait(name string, args ...string) (string, error) {
    cmd := exec.Command(name, args...)

    stdout, err := cmd.StdoutPipe()
    if err != nil {
        return "", err
    }
    stderr, err := cmd.StderrPipe()
    if err != nil {
        return "", err
    }

    err = cmd.Start()
    if err != nil {
        return "", err
    }

    resp, err := io.ReadAll(stdout)  // Blocking here!
    if err != nil {
        return "", err
    }
    errB, err := io.ReadAll(stderr)
    if err != nil {
        //nolint
        return "", nil
    }

    err = cmd.Wait()
    ...
```
Command cannot get a response from `StdoutPipe`, I wrote a simple test and occasionally this happens.
```go
RunCmdAndWait("nerdctl", "run", "--name", "dapr_redis", "-d", "redis:6")
```
But why no data is written to the standard output I do not know yet.. 😅 If the image already exists there is no problem.

shubham1172 commented:
> > @shubham1172 @pravinpushkar According to shubham1172's suggestion I modified `README` and added e2e tests for `containerd` runtime. But the [job](https://github.com/dapr/cli/actions/runs/4524712614/jobs/7968721746) timed out and I am looking for the cause, I suspect that the command `dapr init ...` is invoked in go that does not get the result and make the program waiting. No problem using `dapr init ...` locally
> 
> @shubham1172 @pravinpushkar The problem is that the `utils.RunCmdAndWait` method causes blocking.
> 
> ```go
> func RunCmdAndWait(name string, args ...string) (string, error) {
>     cmd := exec.Command(name, args...)
> 
>     stdout, err := cmd.StdoutPipe()
>     if err != nil {
>         return "", err
>     }
>     stderr, err := cmd.StderrPipe()
>     if err != nil {
>         return "", err
>     }
> 
>     err = cmd.Start()
>     if err != nil {
>         return "", err
>     }
> 
>     resp, err := io.ReadAll(stdout)  // Blocking here!
>     if err != nil {
>         return "", err
>     }
>     errB, err := io.ReadAll(stderr)
>     if err != nil {
>         //nolint
>         return "", nil
>     }
> 
>     err = cmd.Wait()
>     ...
> ```
> 
> Command cannot get a response from `StdoutPipe`, I wrote a simple test and occasionally this happens.
> 
> ```go
> RunCmdAndWait("nerdctl", "run", "--name", "dapr_redis", "-d", "redis:6")
> ```
> 
> But why no data is written to the standard output I do not know yet.. 😅 If the image already exists there is no problem.

Is it because the output is not going to stdout? I see a related issue here https://github.com/containerd/nerdctl/issues/1685

pravinpushkar committed:
Merge branch 'master' into containerd-runtime

shubham1172 commented:
@tianyax did you get a chance to have a look at the issue?

mukundansundar committed:
Merge branch 'master' into containerd-runtime

tianyax commented:
> @tianyax did you get a chance to have a look at the issue?

@shubham1172 I think this problem is caused by `lima/colima` rather than `nerdctl`, I opened a discussion https://github.com/lima-vm/lima/discussions/1454, the maintainer suggested to use `cmd.Stdout = &stdout` to solve this problem, and I responded to it.

tianyax committed:
fixed the issue in e2e tests

Signed-off-by: tianya <tianyax@outlook.com>

pravinpushkar committed:
Merge branch 'master' into containerd-runtime

tianyax commented:
> @tianyax I am seeing that this PR has been moved to draft ... Do you have an ETA for this feature?

@mukundansundar There is a new problem here, using `nerdctl` to quickly create and delete containers will cause errors https://github.com/containerd/nerdctl/discussions/2193. this is occasional in e2e tests (with a few retries have completed the full test case), and it may take some time to fix this bug. so I updated the PR to draft, this feature may need more rigorous and comprehensive testing, otherwise I think it needs to be supported with caution at present😅.. I'm trying to find the cause of the nerdctl error and fix it.

mukundansundar commented:
@tianyax Do you have any updates on this PR?

tianyax commented:
> @tianyax Do you have any updates on this PR?

@mukundansundar Waiting for https://github.com/containerd/containerd/pull/8550
